### PR TITLE
Cache domain handlers when possible

### PR DESCRIPTION
**Pull request by nhunzaker**
This commit changes the way we store snapshots such that they have context into their last operation. This lets us memoize domain handlers. Basically:

1. Is the prior snapshot value the same?
2. Is the prior action status the same?
3. Is the prior action payload the same?

If so, don't invoke the domain handler again!

Testing this in the wild, this has cut extra domain handler invocation for lots of async tasks by about 30%. Pretty neat.

I also believe that we could use this new approach for reporting in the developer tools. We _could_ eventually store each of these snapshots as a diff.

Additionally, if we were to support sub-domains #184, this caching would become more and more powerful as your key paths became more specific. Breaking up complicated state management would get faster.

### CODE REVIEW

File: src/microcosm.js [hunk 1]
```diff
@@ -206,24 +206,44 @@ class Microcosm extends Emitter implements Domain {
...
 
   /**
    * Update the state snapshot for a given action
    */
   updateSnapshot(action: Action) {
-    let next = this.recall(action.parent)
+    // Fall back to creating a snapshot if it does not exist in the event
+    // an action is in progress while a fork is being created
+    let snap = this.snapshots[action.id] || this.createSnapshot(action)
+    let last = this.recall(action.parent)
```
Comment (efatsi) [inline src/microcosm.js]: invalid to go with `let last = snap.last`?

File: src/domain-engine.js [hunk 2]
```diff
@@ -85,17 +85,30 @@ class DomainEngine {
     return domain
   }
 
-  dispatch(state: Object, action: Action): Object {
+  dispatch(action: Action, state: Object, snapshot: Object) {
```
Comment (prayasht) [inline src/domain-engine.js]: This is cool Nate. I'm assuming the `dispatch` method is what actually sends actions (and their payloads) to the store. What exactly does the handler do here?

File: src/microcosm.js (hunk 1 above)
Comment (nhunzaker) [inline src/microcosm.js]: Oh, good idea. Let me see.

Comment (nhunzaker) [inline src/microcosm.js]: It appears that we can not. Though I'd like to figure out how to explain why.

File: src/domain-engine.js (hunk 2 above)
Comment (nhunzaker) [inline src/domain-engine.js]: Yeah. Exactly. The handlers are cached prior invocations of `register()` on domains so that we don't have to call them every time we dispatch an action.

The handlers are what actually modify state, like the value part of register:

```javascript
const Domain = {
  register() {
    return {
      [action]: handler
    }
  }
}
```

File: src/microcosm.js (hunk 1 above)
Comment (nhunzaker) [inline src/microcosm.js]: Ahh. Yes. We can do this. It looks like the `START` action that kicks things off is never getting assigned to a snapshot state.

That's a bummer. Adding START allowed me to avoid some conditional logic, but now it's causing problems. Maybe I can figure out how to remove it.

Comment (nhunzaker) [inline src/microcosm.js]: Okay. So this optimization actually depends on `priorSnapshot.next` and `thisSnapshot.last` to being out of sync. I don't like that this is kind of an obtuse, round-about, linked list. I'm going to see if I can fix that.

Comment (nhunzaker) [inline src/microcosm.js]: I think I have it.

The `last` reference in the snapshot refers to the last state that the revision was calculated with. When we can compare that reference to the current working state to say "did my assumptions change? if so, recalculate this domain handler"

Basically, `last` sort of works as a change diffing tool.

File: src/domain-engine.js [hunk 3]
```diff
@@ -85,17 +85,30 @@ class DomainEngine {
...
+        action.payload !== snapshot.payload ||
+        // or the status is different
+        action.status !== snapshot.status
+      ) {
+        // Yes: recalculate state from the base
+        result = set(result, key, handler.call(source, base, action.payload))
+      } else {
+        // No: use the existing snapshot value (memoizing the domain handler)
+        result = set(result, key, get(snapshot.next, key))
+      }
```
Comment (efatsi) [inline src/domain-engine.js]: this is super clean

File: src/microcosm.js [hunk 4]
```diff
@@ -206,24 +206,38 @@ class Microcosm extends Emitter implements Domain {
...
+    }
   }
 
   /**
    * Update the state snapshot for a given action
    */
   updateSnapshot(action: Action) {
-    let next = this.recall(action.parent)
+    // Fall back to creating a snapshot if it does not exist in the event
+    // an action is in progress while a fork is being created
```
Comment (efatsi) [inline src/microcosm.js]: this comment doesn't seem relevant to the code here

File: src/microcosm.js [hunk 5]
```diff
@@ -206,24 +206,38 @@ class Microcosm extends Emitter implements Domain {
...
+    let snap = this.snapshots[action.id]
+    let last = this.recall(action.parent)
 
     if (this.parent) {
-      next = merge(next, this.parent.recall(action))
+      last = merge(last, this.parent.recall(action))
     }
```
Comment (efatsi) [inline src/microcosm.js]:
```
let last = this.recall(action.parent)
if (this.parent) {
  last = merge(last, this.parent.recall(action))
}
```
should this functionality of reaching up to the parent be moved into the `recall` method? or maybe a `deepRecall` or something. i'm imagining what happens if the parent has a parent.

**Comment by efatsi** [comment]
really cool stuff, the tests are a bit dense and tough to follow since you're using the same function to define an action as well as what the domain calls in response to the action, but overall 👍

File: test/integration/reexecuting-domain-handlers.test.js [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    one.resolve(1)
+
+    expect(addOne).toHaveBeenCalledTimes(1)
+
+    expect(addTwo).toHaveBeenCalledTimes(1)
+
+    expect(repo.state).toEqual({ counterOne: 1, counterTwo: 2 })
+  })
+
+  it('does not call the same domain handler twice when two domains listen to the same action from a different key', function() {
```
Comment (efatsi) [inline test/integration/reexecuting-domain-handlers.test.js]: this test in particular i'm having a tough time following, the "same action from a different key" doesn't line up with what i see below, also why each action is called 3 times and not less is a mystery to me

File: src/microcosm.js (hunk 4 above)
Comment (nhunzaker) [inline src/microcosm.js]: Thank you. A relic from a prior update.

File: src/microcosm.js (hunk 5 above)
Comment (nhunzaker) [inline src/microcosm.js]: Excellent question.

Right now, parents always update their snapshots before their children because they are created first, so they subscribe to history first. This means they get executed first on the list of callbacks whenever history emits update.

So we don't have to do this. But I want to give it a shot. I think it might avoid some edge cases in the future if we start exploring that.

File: test/integration/efficiency.test.js [hunk 7]
```diff
@@ -80,7 +80,7 @@ describe('Efficiency', function() {
     two.resolve()
     one.resolve()
 
-    expect(handler).toHaveBeenCalledTimes(3)
+    expect(handler).toHaveBeenCalledTimes(2)
```
Comment (nhunzaker) [inline test/integration/efficiency.test.js]: Fun little bonus :)

Comment (efatsi) [inline test/integration/efficiency.test.js]: you did say 30% improvement right ;)?

**Comment by nhunzaker** [comment]
@efatsi I've made a new `rebase` method that performs the operations you mentioned regarding `recall`. I've also taken a second pass to improve the clarity of those tests.

File: src/microcosm.js [hunk 8]
```diff
@@ -195,35 +195,57 @@ class Microcosm extends Emitter implements Domain {
...
+   */
+  rebase(action: Action) {
+    let state = this.recall(action.parent)
+
+    if (this.parent) {
+      state = merge(state, this.parent.recall(action))
+    }
+
+    return state
+  }
```
Comment (nhunzaker) [inline src/microcosm.js]: Devtools rely on `recall` to get the state snapshot for an action, so I'm afraid to touch it. We should come up with a public API for getting the snapshot for an action, use that in the dev tools, and revisit this.

File: src/microcosm.js [hunk 9]
```diff
@@ -112,7 +112,7 @@ class Microcosm extends Emitter implements Domain {
 
     this.history = this.parent ? this.parent.history : new History(options)
 
-    this.snapshots = {}
+    this.snapshots = Object.create(this.parent ? this.parent.snapshots : null)
```
Comment (nhunzaker) [inline src/microcosm.js]: This is new. By using prototypal inheritance, forks gain access to prior snapshots without needing to "sync up" as actions update that they couldn't track. Accessing a snapshot prior to when a fork was created causes prototypal access from the parent.

File: src/microcosm.js [hunk 10]
```diff
@@ -193,37 +193,61 @@ class Microcosm extends Emitter implements Domain {
...
 
     if (!action.disabled) {
-      next = this.domains.dispatch(next, action)
+      snap.next = this.domains.dispatch(action, last, snap)
+    } else {
+      snap.next = last
     }
 
-    this.state = this.snapshots[action.id] = next
+    this.snapshots[action.id] = merge(snap, { last, status: action.status, payload: action.payload })
```
Comment (nhunzaker) [inline src/microcosm.js]: Since we share data, we can't use nested mutation. The overhead here appears to be negligible.

File: test/integration/snapshots-in-forks.test.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
+import Microcosm from '../../src/microcosm'
+
+describe('Snapshots', function() {
+  it('forks accommodate actions that were pushed before the parent was forked', function() {
```
Comment (nhunzaker) [inline test/integration/snapshots-in-forks.test.js]: This test specifically covers an edge case I found where:

1. You create a presenter
2. That presenter pushes an action in the setup method
3. A react router route embeds another presenter as a child

This put Microcosm into a state where the fork form the child presenter did not have a snapshot for the parent. The prototypal inheritance update addresses this problem.

File: src/history.js [hunk 12]
```diff
@@ -278,15 +278,16 @@ class History extends Emitter {
     console.assert(action, 'History should never reconcile ' + typeof action)
 
     let focus = action
-
     while (focus) {
-      this._emit('update', focus)
+      if (focus.command !== START) {
```
Comment (nhunzaker) [inline src/history.js]: I think a good next step for History is to figure out how to remove START, BIRTH, and ADD_DOMAIN

File: src/microcosm.js (hunk 9 above)
Comment (prayasht) [inline src/microcosm.js]: That is badass
